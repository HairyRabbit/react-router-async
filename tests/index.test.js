// -*- mode: js-jsx -*-
// -*- coding: utf-8 -*-
// @flow

import React from 'react'
import { Link, Router } from 'react-router-dom'
import createMemoryHistory from 'history/createMemoryHistory'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import Route from '../lib'

const Foo = () => (<h1>foo</h1>)
const Bar = () => (<h1>bar</h1>)
const Baz = () => (<h1>baz</h1>)
const Qux = () => (<h1>qux</h1>)
const HOC = () => Component => Qux

const App = () => (
  <div>
    <Route path='/foo' component={Promise.resolve(Foo)} />
    <Route path='/bar' component={Promise.resolve(Bar)} />
    <Route path='/baz' component={Promise.resolve(Bar)} process={HOC()} />
  </div>
)

const history = createMemoryHistory({
  initialEntries: [ '/', '/foo', '/bar', '/baz' ]
})

test('should render router async', (done) => {
  const component = (
    <Router history={history}>
      <App />
    </Router>
  )

  const wrapper = mount(component)
  const rendered = renderer.create(component)

  let tree

  tree = rendered.toJSON()
  expect(tree).toMatchSnapshot()

  history.replace('/foo')

  setTimeout(() => {
    tree = rendered.toJSON()
    expect(tree).toMatchSnapshot()

    expect(wrapper.find(Foo).exists()).toBe(true)
    expect(wrapper.find(Bar).exists()).toBe(false)
    expect(wrapper.find(Baz).exists()).toBe(false)
    expect(wrapper.find(Qux).exists()).toBe(false)

    history.replace('/bar')

    setTimeout(() => {
      tree = rendered.toJSON()
      expect(tree).toMatchSnapshot()

      expect(wrapper.find(Foo).exists()).toBe(false)
      expect(wrapper.find(Bar).exists()).toBe(true)
      expect(wrapper.find(Baz).exists()).toBe(false)
      expect(wrapper.find(Qux).exists()).toBe(false)

      // Test for process
      history.replace('/baz')

      setTimeout(() => {
        tree = rendered.toJSON()
        expect(tree).toMatchSnapshot()

        expect(wrapper.find(Foo).exists()).toBe(false)
        expect(wrapper.find(Bar).exists()).toBe(false)
        expect(wrapper.find(Baz).exists()).toBe(false)
        expect(wrapper.find(Qux).exists()).toBe(true)

        process.nextTick(done)
      })
    }, 100)
  }, 100)
})
