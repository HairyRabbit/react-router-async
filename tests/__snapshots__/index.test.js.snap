// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render router async 1`] = `<div />`;

exports[`should render router async 2`] = `
<div>
  <h1>
    foo
  </h1>
</div>
`;

exports[`should render router async 3`] = `
<div>
  <h1>
    bar
  </h1>
</div>
`;

exports[`should render router async 4`] = `
<div>
  <h1>
    qux
  </h1>
</div>
`;
